### DISCUSSION TITLE
How to Show hostname or Browser Name like string info in UI

### DISCUSSION THREAD

**Discussion by AjinThankachan** [apache/skywalking#12306] [Q&A]
Hi,
I am using the skywalking 9.7. I am trying to show some metrics like browser name which is s string type in the UI. I know i need to build OAL for the same since its a string its not possible to build or show it using the OAL. Is there any other way to show a string like value.

For example in case of browser monitoring if i send the browser name to skywalking and defined the function in collector like the pagepath which is defined. An i need to show it in the endpoint list is it possibe to do that? I have done the changes in the receiver plugin. But to show in the UI or check the same is reflected what can i do?

Thanks for the guidence.

**Comment by wu-sheng** [comment]
> Is there any other way to show a string like value.

I think I replied to that before. No, it isn't.

> But to show in the UI or check the same is reflected what can i do?

I am not sure what do you mean by this. Endpoint is the page path already. There is no such concept of browser name. It isn't an endpoint.

**Reply by AjinThankachan** [reply]
Ok got it. 
Say for example a person is accessing a website or endpoint in the browser we can record the browsername and hostname and other details along with pagepath.  When send it to skywalking either we need to group it with endpoint or some how we need to show in sywalking UI na. In grafana i think i have seen it. Each endpoint may come from some browser thats why  thought we can group it with endpoint like we are passing the pagepath and show in Skywalking.

Is there any possibility to do with the existing UI? Or do i need to create something outofthe box ? I see in most of the people are interested in this kind of stuff so planned to add the same.

**Reply by wu-sheng** [reply]
I don't know what you mean `seen in grafana`. SkyWalking's Grafana dashboard has nothing different from the native UI.

If you are saying, collecting browser metrics from other platform, that is OTEL or MAL relative. Those ways provide more flexibility to build endpoint name.

**Reply by AjinThankachan** [reply]
Sorry for the confusion. In Grafana faro dashboard i was saying.  Using clint JS itself  i have tried to pass a parameter browserName:chrome, hostName: 0.0.0.0 something like this sort along with the pageurl and other metrics.  

What i am trying to showcase is in the browser UI in skywalking add a widget to show the browser name and hostname if a perticular page is selected. 
![image](https://github.com/apache/skywalking/assets/25667606/c02709c2-a621-4957-b095-c31092a8d151)
Like in the image if i click a page say homepage
![image](https://github.com/apache/skywalking/assets/25667606/60772c5f-d58b-4840-b6f2-042014fdcb13)
In the card layout widget i wish to show the browsername and in the next one hostname. This kind of string data how to fetch and show in dashboard that is my query. I can see only card layout is having the option to display values in a single card. 

If we can do it i think people can identify from which browser or from which host that perticular tranaction is happend and can do some checks based on that. Is it possible to do this?

**Reply by wu-sheng** [reply]
You could use metrics(OTEL or anything we supported) and custom labels to add that.

How to collect OTEL metrics with those labels relies on which agents you are going to use. Once metrics are reported, you could use MAL(rather than OAL) to do further processing. In that way, more labels and processes are provided.

**Reply by AjinThankachan** [reply]
HI @wu-sheng,
Thanks for the replay and sorry for the delay. i was checking what you mentioned. And tried some as well. But i think MAL is also work for Metrics only. If i need to show ip address or host name or any other parameters will that be able to push to skywalking and show the same? Metrics like values or calculations i tried to populate and working fine. But string like values as i mentioned i didnt find any way to do that. Can you give any reference or doc if available to do something like this.

**Reply by wu-sheng** [reply]
I thought I was clear, unless you want to build service, endpoint and instance name by using things from labels, it is impossible to treat string as metrics.

That is not practical.
